Properly handle UTF-8 characters and entities

Renkas opened the pull request:
This pull request will make handling of character entities work properly. As this is something that is dependant on PHPs DOMDocument then basically all that you can do is to add a note in the documentation that input (X)HTML needs to include meta tag with UTF-8 as charset. Otherwise DOMDocument will re-encode all the entities and characters in the document.

I removed the workaround put in to the code as it only made things worse. Added some test cases and added a note to README.md

Partially connected to #172

oliverklee commented on Classes/Emogrifier.php:
```diff
@@ -289,7 +289,7 @@ public function emogrify()
         $this->copyCssWithMediaToStyleNode($cssParts, $xmlDocument);
 
-        return mb_convert_encoding($xmlDocument->saveHTML(), self::ENCODING, 'HTML-ENTITIES');
+        return $xmlDocument->saveHTML();
```
I think the constant ENCODING then is not used anymore either, and we can remove it as well.

oliverklee commented on README.md:
```diff
@@ -63,6 +63,10 @@ After you have set the HTML and CSS, you can call the `emogrify` method to merge
 
+NB! for Emogrifier to properly deal with entities and UTF-8 characters you need to add Content-Type meta tag with UTF-8 charset declaration to your HTML.
+
+    $html = '<html><meta http-equiv="Content-Type" content="text/html; charset=utf-8"><h1>Hello world!</h1></html>';
```
Can't we force Emogrifier to always handle the HTML as UTF-8 instead, thus making the meta tag unnecessary?

oliverklee commented on Tests/Unit/EmogrifierTest.php:
```diff
@@ -96,7 +96,39 @@ public function emogrifyKeepsUtf8Umlauts()
+    /**
+     * @test
+     */
+    public function emogrifyKeepsUtf8UmlautsXhtml()
```
Could you please add an "In" before the "Xhtml"?

oliverklee commented on Tests/Unit/EmogrifierTest.php:
```diff
@@ -96,7 +96,39 @@ public function emogrifyKeepsUtf8Umlauts()
+    {
+        $umlautString = 'Öösel läks õunu täis ämber uhkelt ümber.';
+
+        $html = $this->html4TransitionalDocumentType . '<html><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8"></head><p>' . $umlautString . '</p></html>';
```
Could you please also add a test checking that required HTML entities (e.g. & amp ;) are kept?

oliverklee commented on Tests/Unit/EmogrifierTest.php:
```diff
@@ -96,7 +96,39 @@ public function emogrifyKeepsUtf8Umlauts()
+    /**
+     * @test
+     */
+    public function emogrifyKeepsUtf8UmlautsHtml4()
```
Could you please add an "In" before the "Html"?

oliverklee commented:
Could you please prefix the commit message subject line with "[BUGFIX]" and add the following line to the commit message description (so merging this PR will autoclose that ticket)?

Fixes #172

Renkas commented:
I did some changes.

I removed ENCODING constant
Emogrifier now tries to correct the missing meta tag automatically.
Also added option to disable this.
Added a test for &amp; also - and it passes like it should.

oliverklee commented on Classes/Emogrifier.php:
```diff
@@ -649,7 +661,6 @@ function ($matches) use (&$media) {
     private function createXmlDocument()
     {
         $xmlDocument = new \DOMDocument;
-        $xmlDocument->encoding = self::ENCODING;
```
Doesn't the encoding field have any effect? Or could we avoid having the meta tag if we keep using $xmlDocument->encoding?

oliverklee commented on Classes/Emogrifier.php:
```diff
@@ -681,8 +692,22 @@ private function getUnifiedHtml()
             $bodyWithoutUnprocessableTags = $this->html;
         }
+        
+        if ($this->shouldFixMissingCharsetDeclaration) {
```
Please lets move this block (of the block in the inner or outer if) into a well-named, private method.

oliverklee commented on Classes/Emogrifier.php:
```diff
@@ -681,8 +692,22 @@ private function getUnifiedHtml()
+            if (!stristr($bodyWithoutUnprocessableTags, 'Content-Type')) {
+
+                // We are trying to insert the meta tag to the right spot in the DOM as if we just prepend it to the HTML we will lose attributes set to the HTML tag.
+                if (stristr($bodyWithoutUnprocessableTags, '<head')) {
```
For stristr checks, please always use !== false as the return value isn't always a boolean.

oliverklee commented on Classes/Emogrifier.php:
```diff
@@ -681,8 +692,22 @@ private function getUnifiedHtml()
+                if (stristr($bodyWithoutUnprocessableTags, '<head')) {
+                    $bodyWithoutUnprocessableTags = preg_replace('/<head(.*?)>/i', '<head$1><meta http-equiv="Content-Type" content="text/html; charset=utf-8">', $bodyWithoutUnprocessableTags);
+                } elseif (stristr($bodyWithoutUnprocessableTags, '<html')) {
+                    $bodyWithoutUnprocessableTags = preg_replace('/<html(.*?)>/i', '<html$1><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8"></head>', $bodyWithoutUnprocessableTags);
```
We have duplicate code here (at least the meta tag). Please move it to a local variable.

oliverklee commented on README.md:
```diff
@@ -63,6 +63,11 @@ After you have set the HTML and CSS, you can call the `emogrify` method to merge
     $mergedHtml = $emogrifier->emogrify();
 
+For Emogrifier to properly deal with entities and UTF-8 characters you need to add Content-Type meta tag with UTF-8 charset
+declaration to your HTML. If you don't provide this tag in your HTML then Emogrifier adds it automatically.
```
Could you please reformat the lines in the README to be under 80 characters? This will avoid wrapping on GitHub.

oliverklee replied:
I think there should be a comma after "HTML".

oliverklee commented on README.md:
```diff
@@ -63,6 +63,11 @@ After you have set the HTML and CSS, you can call the `emogrify` method to merge
 
     $mergedHtml = $emogrifier->emogrify();
 
+For Emogrifier to properly deal with entities and UTF-8 characters you need to add Content-Type meta tag with UTF-8 charset
```
I think there should be a comma after "characters".

oliverklee commented on Classes/Emogrifier.php:
```diff
@@ -323,6 +325,16 @@ public function disableInvisibleNodeRemoval()
+     *
+     * @return void
+     */
+    public function disableCharsetFix()
```
We need unit tests for this method.

(Or, I'd prefer to always have this new behavior, and then we don't need that switch.)

oliverklee commented:
Could you please squash your commits?

oliverklee commented:
Could you also please rebase from master? I've merged some PRs.

oliverklee commented on Tests/Unit/EmogrifierTest.php:
```diff
@@ -108,9 +108,41 @@ public function emogrifyKeepsUtf8Umlauts()
+        $html = $this->xhtml1StrictDocumentType . '<html style="padding: 1px;"><p>' . $umlautString . '</p></html>';
+        $this->subject->setHtml($html);
+
+        $this->assertContains(
```
Could you please change this to be a static call: self::assertContains

oliverklee commented on Tests/Unit/EmogrifierTest.php:
```diff
@@ -108,9 +108,41 @@ public function emogrifyKeepsUtf8Umlauts()
+        $html = $this->html4TransitionalDocumentType . '<html><p>' . $umlautString . '</p></html>';
+        $this->subject->setHtml($html);
+
+        $this->assertContains(
```
Could you please change this to be a static call: self::assertContains

oliverklee commented:
Hi @Renkas, I've taken your PR, polished it an merged it as #189. Thanks a lot for contributing!
